### PR TITLE
HBase port from Paris Hackathon

### PR DESCRIPTION
**Pull request by mleich**
So this is the initial version of HBase for ozone.
Sample code how to read from HBase is provided in pact-examples

I merged this with dimalabs:master.
The relevant commits are 33d6a1a and 39664be
The all previous commits are from the cloudera support branch, everything after those two is the result of merging.

Get back to me if this is too messy, I could also just copy/paste the HBase stuff into a fresh branch.
The current diff to dimalabs:master looks find to me, though.

### CODE REVIEW

File: pact/pact-examples/pom.xml [hunk 1]
```diff
@@ -40,6 +47,29 @@
 				<version>2.4</version>
 				<executions>
 				
+					<!-- HBaseExample -->
```
Comment (StephanEwen) [inline pact/pact-examples/pom.xml]: Not every example in pact-examples needs to be packaged as a jar. If the code is more for illustration issues and does not really constitute a program that anyone would execute, I think we should not package it into a jar. Makes the set of examples programs kind of crowded.

File: pact/pact-hbase/pom.xml [hunk 2]
```diff
@@ -0,0 +1,39 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
+  <modelVersion>4.0.0</modelVersion>
+  <parent>
+    <artifactId>pact</artifactId>
+    <groupId>eu.stratosphere</groupId>
+    <version>0.2</version>
```
Comment (StephanEwen) [inline pact/pact-hbase/pom.xml]: please update to version 0.2-ozone

File: pact/pact-hbase/src/main/java/eu/stratosphere/pact/common/contract/HBaseDataSink.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+import java.util.Random;
+
+import eu.stratosphere.pact.common.io.GenericTableOutputFormat;
+import eu.stratosphere.pact.generic.contract.Contract;
+
+
+/**
+ *
+ *
+ * @author Stephan Ewen (stephan.ewen@tu-berlin.de)
```
Comment (StephanEwen) [inline pact/pact-hbase/src/main/java/eu/stratosphere/pact/common/contract/HBaseDataSink.java]: Feel free to remove author tags. We don't want code ownership, everyone should feel free to dive into whatever component.

File: pact/pact-hbase/src/main/java/eu/stratosphere/pact/common/contract/HBaseDataSink.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+		
+		// generate a random unique identifier string
+		final Random rnd = new Random();
+		final StringBuilder bld = new StringBuilder();
+		for (int i = 0; i < IDENTIFYIER_LEN; i++) {
+			bld.append((char) (rnd.nextInt(26) + 'a'));
+		}
+		
+		setParameter(GenericTableOutputFormat.JOB_ID_KEY, bld.toString());
+		setParameter(GenericTableOutputFormat.JOB_ID_KEY, rnd.nextInt());
```
Comment (StephanEwen) [inline pact/pact-hbase/src/main/java/eu/stratosphere/pact/common/contract/HBaseDataSink.java:28]: This does not look right. The value in the previous line is plainly overwritten.

Comment (StephanEwen) [inline pact/pact-hbase/src/main/java/eu/stratosphere/pact/common/contract/HBaseDataSink.java:28]: Looking into GenericTableOutputFormat, it seems one of them should be the JT_ID_KEY.

File: pact/pact-hbase/src/main/java/eu/stratosphere/pact/common/io/TableOutputFormat.java [hunk 5]
```diff
@@ -0,0 +1,280 @@
+//package eu.stratosphere.pact.common.io;
```
Comment (StephanEwen) [inline pact/pact-hbase/src/main/java/eu/stratosphere/pact/common/io/TableOutputFormat.java]: Should we include files that are fully commented out? Does it even compile with the package declaration commented out.

IS this class currently defunct?

File: pact/pact-hbase/src/test/java/eu/stratosphere/pact/common/io/test/ShortIMRAccess.java [hunk 6]
```diff
@@ -0,0 +1,60 @@
+/**
```
Comment (StephanEwen) [inline pact/pact-hbase/src/test/java/eu/stratosphere/pact/common/io/test/ShortIMRAccess.java]: Remove test that is IMR specific.

File: pact/pact-hbase/src/test/java/eu/stratosphere/pact/common/io/test/standAloneRead.java [hunk 7]
```diff
@@ -0,0 +1,96 @@
+package eu.stratosphere.pact.common.io.test;
```
Comment (StephanEwen) [inline pact/pact-hbase/src/test/java/eu/stratosphere/pact/common/io/test/standAloneRead.java]: Please format the file.

File: pom.xml [hunk 8]
```diff
@@ -152,6 +152,17 @@
 			<scope>compile</scope>
 		</dependency>
 		-->
+
```
Comment (StephanEwen) [inline pom.xml]: Is this dependency introduced by us? We should not have two logging frameworks side by side

Please remove if possible and use the already included Apache Commons logging (which is also a facade to multiple logging frameworks).

Comment (mleich) [inline pom.xml]: Without this HBase will not work. I get the following error message:

``` java
SLF4J: slf4j-api 1.6.x (or later) is incompatible with this binding.
SLF4J: Your binding is version 1.5.5 or earlier.
SLF4J: Upgrade your binding to version 1.6.x. or 2.0.x
ERROR: The program execution failed: eu.stratosphere.nephele.executiongraph.GraphConversionException: java.lang.RuntimeException: The user defined 'configure()' method caused an error: org.slf4j.impl.StaticLoggerBinder.getSingleton()Lorg/slf4j/impl/StaticLoggerBinder;
    at eu.stratosphere.pact.runtime.task.DataSourceTask.initInputFormat(DataSourceTask.java:296)
    at eu.stratosphere.pact.runtime.task.DataSourceTask.registerInputOutput(DataSourceTask.java:95)
    at eu.stratosphere.nephele.execution.RuntimeEnvironment.<init>(RuntimeEnvironment.java:196)
    at eu.stratosphere.nephele.executiongraph.ExecutionGroupVertex.<init>(ExecutionGroupVertex.java:240)
    at eu.stratosphere.nephele.executiongraph.ExecutionGraph.createVertex(ExecutionGraph.java:515)
    at eu.stratosphere.nephele.executiongraph.ExecutionGraph.constructExecutionGraph(ExecutionGraph.java:286)
    at eu.stratosphere.nephele.executiongraph.ExecutionGraph.<init>(ExecutionGraph.java:177)
    at eu.stratosphere.nephele.jobmanager.JobManager.submitJob(JobManager.java:535)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:606)
    at eu.stratosphere.nephele.ipc.RPC$Server.call(RPC.java:417)
    at eu.stratosphere.nephele.ipc.Server$Handler.run(Server.java:946)
Caused by: java.lang.NoSuchMethodError: org.slf4j.impl.StaticLoggerBinder.getSingleton()Lorg/slf4j/impl/StaticLoggerBinder;
    at org.slf4j.LoggerFactory.bind(LoggerFactory.java:121)
    at org.slf4j.LoggerFactory.performInitialization(LoggerFactory.java:111)
    at org.slf4j.LoggerFactory.getILoggerFactory(LoggerFactory.java:268)
    at org.slf4j.LoggerFactory.getLogger(LoggerFactory.java:241)
    at org.slf4j.LoggerFactory.getLogger(LoggerFactory.java:254)
    at org.apache.hadoop.security.authentication.util.KerberosName.<clinit>(KerberosName.java:42)
    at org.apache.hadoop.security.UserGroupInformation.ensureInitialized(UserGroupInformation.java:214)
    at org.apache.hadoop.security.UserGroupInformation.isSecurityEnabled(UserGroupInformation.java:277)
    at org.apache.hadoop.security.UserGroupInformation.getLoginUser(UserGroupInformation.java:668)
    at org.apache.hadoop.security.UserGroupInformation.getCurrentUser(UserGroupInformation.java:573)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:606)
    at org.apache.hadoop.hbase.util.Methods.call(Methods.java:37)
    at org.apache.hadoop.hbase.security.User.call(User.java:586)
    at org.apache.hadoop.hbase.security.User.callStatic(User.java:576)
    at org.apache.hadoop.hbase.security.User.access$400(User.java:50)
    at org.apache.hadoop.hbase.security.User$SecureHadoopUser.<init>(User.java:393)
    at org.apache.hadoop.hbase.security.User$SecureHadoopUser.<init>(User.java:388)
    at org.apache.hadoop.hbase.security.User.getCurrent(User.java:139)
    at org.apache.hadoop.hbase.client.HConnectionManager$HConnectionKey.<init>(HConnectionManager.java:416)
    at org.apache.hadoop.hbase.client.HConnectionManager.getConnection(HConnectionManager.java:179)
    at org.apache.hadoop.hbase.client.HTable.<init>(HTable.java:192)
    at org.apache.hadoop.hbase.client.HTable.<init>(HTable.java:170)
    at eu.stratosphere.pact.common.io.TableInputFormat.createTable(TableInputFormat.java:172)
    at eu.stratosphere.pact.example.hbase.HBaseReadExample$MyTableInputFormat.createTable(HBaseReadExample.java:60)
    at eu.stratosphere.pact.common.io.TableInputFormat.configure(TableInputFormat.java:94)
    at eu.stratosphere.pact.runtime.task.DataSourceTask.initInputFormat(DataSourceTask.java:293)
    ... 13 more

    at eu.stratosphere.nephele.executiongraph.ExecutionGraph.createVertex(ExecutionGraph.java:521)
    at eu.stratosphere.nephele.executiongraph.ExecutionGraph.constructExecutionGraph(ExecutionGraph.java:286)
    at eu.stratosphere.nephele.executiongraph.ExecutionGraph.<init>(ExecutionGraph.java:177)
    at eu.stratosphere.nephele.jobmanager.JobManager.submitJob(JobManager.java:535)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:606)
    at eu.stratosphere.nephele.ipc.RPC$Server.call(RPC.java:417)
    at eu.stratosphere.nephele.ipc.Server$Handler.run(Server.java:946)
```

File: pact/pact-examples/pom.xml (hunk 1 above)
Comment (mleich) [inline pact/pact-examples/pom.xml]: I'd suggest to comment that one out. The example comes in handy for tests.

**Comment by StephanEwen** [comment]
Manually merged and fixed in 8f7484000ae40ae1b5c68e2d21b9e5a44d984234

**Pull request closed without merging**